toss/fitness/fitness_function_utils.py

```diff
@@ -143,25 +117,140 @@ def compute_space_coverage(number_of_spacecrafts: int, spin_axis: np.ndarray, sp
+    r_points, theta_points, phi_points = cart2sphere(rotated_positions[0,:], rotated_positions[1,:], rotated_positions[2,:])
+
+    # Remove points outside measurement zone (i.e outside outer-bounding sphere)
+    index_feasible_positions = np.where(r_points <= radius_max) # Addition of noise to cover approximation error
+    r_points = r_points[index_feasible_positions]
+    theta_points = theta_points[index_feasible_positions]
+    phi_points = phi_points[index_feasible_positions]
+
+    # Remove points inside safety-radius (i.e inside inner-bounding sphere)
+    index_feasible_positions = np.where(r_points >= radius_min) # Addition of noise to cover approximation error
+    r_points = r_points[index_feasible_positions]
+    theta_points = theta_points[index_feasible_positions]
+    phi_points = phi_points[index_feasible_positions]
+
+    # Update and return boolean tensory given information on the new trajectory
+    if len(r_points)==0 or len(theta_points)==0 or len(phi_points)==0:
+        return bool_tensor
+    else: 
+        # Find the indices of the closest values in the meshgrid for each point using broadcasting
+        i = np.argmin(np.abs(r[:, np.newaxis] - r_points), axis=0) # indices along r axis
+        j = np.argmin(np.abs(theta[:, np.newaxis] - theta_points), axis=0) # indices along theta axis
+        k = np.argmin(np.abs(phi[:, np.newaxis] - phi_points), axis=0) # indices along phi axis
+
+        # Update tensor with information on new trajectory using advanced indexing
+        bool_tensor[i, j, k] = True
+        return bool_tensor
+
+
+def get_spherical_tensor_grid(time_step: int, radius_min: float, radius_max: float, max_velocity_scaling_factor: float) -> Union[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
+    """
+    Generates a number of points in each spherical axis, which together defines
+    a spherical tensor grid that satisfies the Courant–Friedrichs–Lewy condition.
+    The function also generates a corresponding boolean tensor initialized by False-values
+    indicating none of the defined points have been visited by a spacecraft.
+
+    Args:
+        timesteps (int): Fixed measurement period (i.e time step along trajectory).
+        radius_min (float): Inner radius of spherical grid, typically radius_inner_bounding_sphere.
+        radius_max (float): Outer radius of spherical grid, typically radius_outer_bounding_sphere.
+        max_velocity_scaling_factor (float): Scales the magnitude of the fixed-valued maximal velocity and therefore also the grid spacing.
+
+    Returns:
+        r (np.ndarray): Array of r coordinates for each point defined on the spherical tensor.
+        theta (np.ndarray): Array of theta coordinates for each point defined on the spherical tensor.
+        phi (np.ndarray): Array of phi coordinates for each point defined on the spherical tensor.
+        bool_tensor (np.ndarray): Boolean array corresponding to each point defined on the spherical tensor.
+    """
+
```

Comment:
some asserts on the input variables wouldn't hurt

`assert timesteps > 0` etc

Reply:
Fixed.